### PR TITLE
ODM: counterparts for two new CDS @ODM annotations 

### PR DESCRIPTION
**Pull request by ralfhandl**
Added
- `ODM.oid`: same name in CDS
- `ODM.entityName`: same name in CDS

Fixed
- removed accidental nullability of `ODM.root` and `ODM.codeList`

### CODE REVIEW

File: vocabularies/ODM.xml [hunk 1]
```diff
@@ -33,17 +33,35 @@
...
+        <Annotation Term="Core.Description" String="The named field is an OID (ODM identifier), and it uniquely identifies an ODM root entity in a customer landscape." />
+        <Annotation Term="Core.LongDescription">
+          <String>The technical type can be UUID or String (any length), and the value must be stable and unique at least within type and landscape (better globally). 
+An entity may have other IDs, i.e. a local ID and other alternative IDs. This is the ID used for ODM.</String>
+        </Annotation>
+      </Term>
+
+      <Term Name="entityName" Type="Edm.String" Nullable="false" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="This annotation points to an ODM entity as a general concept, not a concrete version thereof" />
```
Comment (HeikoTheissen) [inline vocabularies/ODM.xml:58]:
```suggestion
        <Annotation Term="Core.Description" String="Name of an ODM entity as a general concept, not a concrete version thereof" />
```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

**Review by HeikoTheissen — APPROVED** [review]

**Comment by ralfhandl** [comment]
@swennemers can we merge this PR?

**Comment by swennemers** [comment]
@ralfhandl wow, you are so fast. So far the usage of the @ODM vocabulary is not aligned with the stakeholders. Not sure if that is enough to already merge the PR.

**Comment by ralfhandl** [comment]
@swennemers Then let's wait for the stakeholder alignment.

Minor note: in CAP this is the `@ODM` vocabulary - case matters, annotations starting with a lowercase letter (for example `@cds`) are treated as "internal annotations" and are not propagated to OData $metadata or OpenAPI, only annotations starting with an uppercase letter (`@Core`, `@Capabilities`, ...) are propagated to OData $metadata.

File: vocabularies/ODM.json [hunk 2]
```diff
@@ -52,14 +51,32 @@
...
+        },
+        "oid": {
+            "$Kind": "Term",
+            "$Type": "Edm.PropertyPath",
+            "$AppliesTo": [
+                "EntityType"
+            ],
+            "@Common.Experimental": true,
+            "@Core.Description": "The named field is an OID (ODM identifier), and it uniquely identifies an ODM root entity in a customer landscape.",
+            "@Core.LongDescription": "The technical type can be UUID or String (any length), and the value must be stable and unique at least within type and landscape (better globally). \nAn entity may have other IDs, i.e. a local ID and other alternative IDs. This is the ID used for ODM."
```
Comment (swennemers) [inline vocabularies/ODM.json:70]:
```suggestion
            "@Core.LongDescription": "The field is an alternative key to the entity. The technical type can be UUID or String (typically length 128), and the value must be stable and unique at least within type and landscape (better globally). \nAn entity may have other IDs, i.e. a local ID and other alternative IDs. "
```

File: vocabularies/ODM.json [hunk 3]
```diff
@@ -52,14 +51,32 @@
...
+        },
+        "entityName": {
+            "$Kind": "Term",
+            "$AppliesTo": [
+                "EntityType"
+            ],
+            "@Common.Experimental": true,
+            "@Core.Description": "Name of an ODM entity as a general concept, not a concrete version thereof",
+            "@Core.LongDescription": "This implies that the annotation cannot provide any contract about properties being contained in a model.\nThe annotated entity type thus is _not_ an implementation of the ODM entity, it only represents this ODM entity."
```
Comment (swennemers) [inline vocabularies/ODM.json:79]: @ralfhandl I am not sure about this description. With recent ODM additions, there are multiple representations of a single entity...

**Review by swennemers — CHANGES_REQUESTED** [review]
Thanks a lot for the update to the PR. @everyone reading this PR in the future, these annotations have now their direct representation in CDS: @ODM.entityName and @ODM.oid.

File: vocabularies/ODM.json (hunk 3 above)
Comment (swennemers) [reply vocabularies/ODM.json:79]: not sure where you found this specific description and long description... But I missed that the xml is the source of truth of the json and md files

File: vocabularies/ODM.xml [hunk 4]
```diff
@@ -33,17 +33,35 @@
...
         <Annotation Term="Common.Experimental" />
         <Annotation Term="Core.Description" String="The annotated entity set contains root entities" />
         <Annotation Term="Core.LongDescription" String="Root Entities have an independent lifetime and are reachable as top level resources in APIs. These are usually the key entities of a domain, sometimes also called 'business objects'. They are called aggregate roots in DDD terminology." />
       </Term>
 
+      <Term Name="oid" Type="Edm.PropertyPath" Nullable="false" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="The named field is an OID (ODM identifier), and it uniquely identifies an ODM root entity in a customer landscape." />
+        <Annotation Term="Core.LongDescription">
+          <String>The technical type can be UUID or String (any length), and the value must be stable and unique at least within type and landscape (better globally).
```
Comment (swennemers) [inline vocabularies/ODM.xml:51]:
```suggestion
          <String>The technical type can be UUID or String (typical length 128), and the value must be stable and unique at least within type and landscape (better globally). 
```

Comment (swennemers) [reply vocabularies/ODM.xml:51]: I understand that wasn't part of the CDS request, but there was a link to an SAP internal ADR which isn't available outside.

File: vocabularies/ODM.xml [hunk 5]
```diff
@@ -33,17 +33,35 @@
...
         <Annotation Term="Core.Description" String="The annotated entity set contains root entities" />
         <Annotation Term="Core.LongDescription" String="Root Entities have an independent lifetime and are reachable as top level resources in APIs. These are usually the key entities of a domain, sometimes also called 'business objects'. They are called aggregate roots in DDD terminology." />
       </Term>
 
+      <Term Name="oid" Type="Edm.PropertyPath" Nullable="false" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="The named field is an OID (ODM identifier), and it uniquely identifies an ODM root entity in a customer landscape." />
+        <Annotation Term="Core.LongDescription">
+          <String>The technical type can be UUID or String (any length), and the value must be stable and unique at least within type and landscape (better globally). 
+An entity may have other IDs, i.e. a local ID and other alternative IDs. This is the ID used for ODM.</String>
```
Comment (swennemers) [inline vocabularies/ODM.xml:52]:
```suggestion
An entity may have other IDs, i.e. a local ID and other alternative IDs. </String>
```
"This is the ID used for ODM" should be obvious from being an ODM ID. It doesn't add more to it, does it? Therefore I would remove it.

File: vocabularies/ODM.xml [hunk 6]
```diff
@@ -33,17 +33,35 @@
...
+        </Annotation>
+      </Term>
+
+      <Term Name="entityName" Type="Edm.String" Nullable="false" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="Name of an ODM entity as a general concept, not a concrete version thereof" />
+        <Annotation Term="Core.LongDescription">
+          <String>This implies that the annotation cannot provide any contract about properties being contained in a model.
+The annotated entity type thus is _not_ an implementation of the ODM entity, it only represents this ODM entity.</String>
```
Comment (swennemers) [inline vocabularies/ODM.xml:61]:
```suggestion
          <String>The annotated OData entity is one of many representations of the ODM entity. Annotating the OData entity with the annotation helps consumers finding APIs that process or expose the same entity.</String>
```
A new proposal trying to express it a bit more positively.

**Review by swennemers — CHANGES_REQUESTED** [review]

**Review by swennemers — APPROVED** [review]

File: vocabularies/ODM.xml [hunk 7]
```diff
@@ -33,17 +33,34 @@
...
+          <String>The technical type can be UUID or String (typical length 128), and the value must be stable and unique at least within type and landscape (better globally). 
+An entity may have other IDs, i.e. a local ID and other alternative IDs. </String>
+        </Annotation>
+      </Term>
+
+      <Term Name="entityName" Type="Edm.String" Nullable="false" AppliesTo="EntityType">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="Name of an ODM entity as a general concept, not a concrete version thereof" />
+        <Annotation Term="Core.LongDescription">
+          <String>The annotated OData entity is one of many representations of the ODM entity. Annotating the OData entity with the annotation helps consumers finding APIs that process or expose the same entity.</String>
```
Comment (HeikoTheissen) [inline vocabularies/ODM.xml:60]:
```suggestion
          <String>The annotated OData entity is one of many representations of the ODM entity. Annotating the OData entity with this term helps consumers find APIs that process or expose the same entity.</String>
```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: vocabularies/ODM.xml (hunk 7 above)
Comment (ralfhandl) [reply vocabularies/ODM.xml:60]: Will open a new PR for this change

**Pull request merged**